examples/nat_scapy_tests.py

```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/python
+# coding: latin-1
+
+"""
+nat_scapy_tests.py: Rudimentary servers/clients for use in testing of the NAT.
+Jonny Shipton, Cambridge University Computer Lab, July 2016
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+
+This script is used in nat_topo.py in the testing of the NAT implementation.
+The main reason for using Scapy is to be able to do things that may not be
+possible with the OS network stack, such as sending packets during the TIME_WAIT state.
+"""
+
+from scapy.all import *
+from scapy.layers.inet import *
+
+# Values that need to be kept in sync with those in the wiring config:
+# NOTE that all times are in seconds
+tcp_inactivity_timeout = 20
+tcp_time_wait_duration = 10
+udp_inactivity_timeout= 200
+
+# The IP addresses of the nodes we use
+serverhost = "10.0.0.4"
+nathost = "10.0.0.3"
+clienthost = "192.168.1.10"
+
+# Templates for iptables rules
+iptables_rule_fmt = "INPUT -p tcp --sport %d --dport %d -j DROP"
+iptables_add_rule_fmt = "iptables -A %s"
+iptables_remove_rule_fmt = "iptables -D %s"
+
+# Data to be sent through the NAT
+server_data = "SERVER DATA"
+client_data = "CLIENT DATA"
+
+def run_server(serverport=12012, natport=35001):
+    "This should be run on the external host."
+
+    # Set up iptables rule to allow us exclusive access on our port
+    iptables_rule = iptables_rule_fmt % (natport, serverport)
+    print "  server> $ %s" % (iptables_add_rule_fmt % iptables_rule)
+    subprocess.check_call(iptables_add_rule_fmt % iptables_rule, shell=True)
+
+    # We use the same IP layer and filter each time
+    ip = IP(src=serverhost, dst=nathost)
+    filter = "tcp and host %s and port %d" % (nathost, serverport)
+
+    # Wait for the connection to be initialised from the client
+    print "Waiting for connection from client"
+    syn = sniff(count=1, filter=filter)[0]
```

Comment:
It would be prudent here to check that the Syn flag in the received packet is indeed set.